games.d/fortune.d/strfile.8
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.ie t \{\
.	ds lq ``
.	ds rq ''
.\}
.el \{\
.	ds lq "
.	ds rq "
.\}
.TH STRFILE 8 "31 Apr, 1987"
.UC 4
.SH NAME
strfile, unstr \- create a random access file for storing strings
.SH SYNOPSIS
.B strfile
[
.B \-
] [
.B \-svc\fIC\fP
] [
.B \-oir
]
sourcefile
[ datafile ]
.sp
.B unstr
[
.B \-o
]
datafile[.dat] [ outfile ]
.SH DESCRIPTION
.I strfile
is designed to take a file which contains a set of strings
and create a data file which contains those strings,
along with a seek pointer table
to the beginning of each.
This allows random access of the strings.
.PP
The source file contains strings
separated by lines starting with \*(lq%%\*(rq or \*(lq%\-\*(rq
Anything following these starting characters on the line
will be ignored,
so comments can be placed on these lines.
A \*(lq%%\*(rq simply separates strings;
a \*(lq%\-\*(rq separates not only strings but sections.
A file can have up to four sections,
.I i.e. ,
up to three delimiters.
This can be used in a program-defined way.
.PP
The data file,
if not specified on the command line,
is named
.IB sourcefile .out ,
and contains a header,
which describes the contents of the data file,
the seek pointers to the beginning of each string,
and the strings themselves,
terminated by null bytes.
.PP
The format of the header is:
.sp
.nf
.ta
.ta 8n +\w'unsigned long\ \ 'u +\w'str_delims[MAXDELIMS];\ 'u
# define	MAXDELIMS	3

# define	STR_RANDOM	0x1
# define	STR_ORDERED	0x2

typedef struct {
	unsigned long	str_numstr;	/* # of strings in the file */
	unsigned long	str_longlen;	/* length of longest string */
	unsigned long	str_shortlen;	/* length of shortest string */
	long	str_delims[MAXDELIMS];	/* delimiter markings */
	off_t	str_dpos[MAXDELIMS];	/* delimiter positions */
	short	str_flags;	/* bit field for flags */
} STRFILE;
.ev
.fi
.PP
The values in
.B str_delims
are the indices of the first string
which follow each \*(lq%\-\*(rq in the file.
The field
.B str_flags
fill have the bit
.SM STR_RANDOM
set if the
.B \-r
flag was specified,
or
.SM STR_ORDERED
if the
.B \-o
flag was specified.
.PP
The options are:
.TP
.B \-
Give a usage summary.
.TP
.B \-s
Run silently;
don't give the summary of data at the end.
.TP
.B \-v
Verbose (default).
.TP
.B \-o
Order the strings in alphabetical order.
The strings will be stored in the same order in the data file
as they were in the source,
but the seek pointer table will be sorted in alphabetical order
of the strings the point to.
Any
.I initial
non-alphanumeric characters are ignored.
This sets the
.SM STR_ORDERED
bit in the
.B str_flags
field of the header.
.TP
.B \-i
Ignore case when ordering.
.TP
.B \-r
Randomize the order of the seek pointers in the table.
The strings will be stored in the same order in the data file
as they were in the source,
but the seek pointer table will be randomized.
This sets the
.SM STR_RANDOM
bit in the
.B str_flags
field of the header.
.TP
.BI \-c C
Change the delimiting character from \*(lq%\*(rq to
.RI \\*(lq C \\*(rq,
making the delimiting lines start with
.RI \\*(lq CC \\*(rq
or
.RI \\*(lq C \-\\*(rq.
.PP
The purpose of
.I unstr
is to undo the work of
.I strfile .
It primarily exists as an emergency backup
in case you accidentally delete your source file,
but still have your data file around.
It reads data files and creates a corresponding output file.
If you don't want \*(lq%\*(rq as your delimiting character,
you can use the
.BI \-c C
option to change it.
.PP
.I unstr
will normally print out the strings
in the order they are in the data file.
If you give it the
.B \-o
option,
it will print them out in the header order,
which could be different if the file was
randomized or ordered when created.
Using this,
you can created sorted versions of your input file
by using
.B \-o
when you run
.IR strfile , 
and then using
.B "unstr \-o"
to dump them out in the table order.
.SH "SEE ALSO"
fortune(6)


games.d/fortune.d/strfile.c
# include	<stdio.h>
# include	<ctype.h>
# include	<sys/types.h>
# include	"strfile.h"

/*
 *	This program takes a file composed of strings seperated by
 * lines starting with two consecutive delimiting character (default
 * character is '%') and creates another file which consists of a table
 * describing the file (structure from "strfile.h"), a table of seek
 * pointers to the start of the strings, and the strings, each terinated
 * by a null byte.  Usage:
 *
 *	% strfile [ - ] [ -cC ] [ -sv ] [ -oir ] sourcefile [ datafile ]
 *
 *	- - Give a usage summary useful for jogging the memory
 *	c - Change delimiting character from '%' to 'C'
 *	s - Silent.  Give no summary of data processed at the end of
 *	    the run.
 *	v - Verbose.  Give summary of data processed.  (Default)
 *	o - order the strings in alphabetic order
 *	i - if ordering, ignore case 
 *	r - randomize the order of the strings
 *
 *		Ken Arnold	Sept. 7, 1978 --
 *
 *	Added method to indicate dividers.  A "%-" will cause the address
 * to be added to the structure in one of the pointer elements.
 *
 *		Ken Arnold	Nov., 1984 --
 *
 *	Added ordering options.
 */

# define	TRUE	1
# define	FALSE	0

# define	DELIM_CH	'-'

typedef struct {
	char	first;
	off_t	pos;
} STR;

char	*Infile		= NULL,		/* input file name */
	Outfile[100]	= "",		/* output file name */
	Delimch		= '%',		/* delimiting character */
	*Usage[]	= {		/* usage summary */
       "usage:	strfile [ - ] [ -cC ] [ -sv ] [ -oir ] inputfile [ datafile ]",
       "	- - Give this usage summary",
       "	c - Replace delimiting character with 'C'",
       "	s - Silent.  Give no summary",
       "	v - Verbose.  Give summary.  (default)",
       "	o - order strings alphabetically",
       "	i - ignore case in ordering",
       "	r - randomize the order of the strings",
       "	Default \"datafile\" is inputfile.dat",
	NULL
	};

int	Sflag		= FALSE;	/* silent run flag */
int	Oflag		= FALSE;	/* ordering flag */
int	Iflag		= FALSE;	/* ignore case flag */
int	Rflag		= FALSE;	/* randomize order flag */
int	Delim		= 0;		/* current delimiter number */

off_t	*Seekpts;

FILE	*Sort_1, *Sort_2;		/* pointers for sorting */

STRFILE	Tbl;				/* statistics table */

STR	*Firstch;			/* first chars of each string */

char	*fgets(), *malloc(), *strcpy(), *strcat();

main(ac, av)
int	ac;
char	**av;
{
	register char		*sp, dc;
	register off_t		*lp;
	register unsigned int	curseek;	/* number of strings */
	register off_t		*seekpts, li;	/* table of seek pointers */
	register FILE		*inf, *outf;
	register int		first;
	register char		*nsp;
	register STR		*fp;
	static char		string[257];

	getargs(ac, av);		/* evalute arguments */

	/*
	 * initial counting of input file
	 */

	dc = Delimch;
	if ((inf = fopen(Infile, "r")) == NULL) {
		perror(Infile);
		exit(-1);
	}
	for (curseek = 0; (sp = fgets(string, 256, inf)) != NULL; )
		if (*sp++ == dc && (*sp == dc || *sp == DELIM_CH))
			curseek++;
	curseek++;

	/*
	 * save space at begginning of file for tables
	 */

	if ((outf = fopen(Outfile, "w")) == NULL) {
		perror(Outfile);
		exit(-1);
	}

	/*
	 * Allocate space for the pointers, adding one to the end so the
	 * length of the final string can be calculated.
	 */
	++curseek;
	seekpts = (off_t *) malloc(sizeof *seekpts * curseek);	/* NOSTRICT */
	if (seekpts == NULL) {
		perror("calloc");
		exit(-1);
	}
	if (Oflag) {
		Firstch = (STR *) malloc(sizeof *Firstch * curseek);
		if (Firstch == NULL) {
			perror("calloc");
			exit(-1);
		}
	}

	(void) fseek(outf, (off_t) (sizeof Tbl + sizeof *seekpts * curseek), 0);
	(void) fseek(inf, (off_t) 0, 0);	/* goto start of input */

	/*
	 * write the strings onto the file
	 */

	Tbl.str_longlen = 0;
	Tbl.str_shortlen = (unsigned int) 0xffffffff;
	lp = seekpts;
	first = Oflag;
	*seekpts = ftell(outf);
	fp = Firstch;
	do {
		sp = fgets(string, 256, inf);
		if (sp == NULL ||
		    (*sp == dc && (sp[1] == dc || sp[1] == DELIM_CH))) {
			putc('\0', outf);
			*++lp = ftell(outf);
			li = ftell(outf) - lp[-1] - 1;
			if (Tbl.str_longlen < li)
				Tbl.str_longlen = li;
			if (Tbl.str_shortlen > li)
				Tbl.str_shortlen = li;
			if (sp && sp[1] == DELIM_CH && Delim < MAXDELIMS) {
				Tbl.str_dpos[Delim] = *lp;
				Tbl.str_delims[Delim++] = lp - seekpts;
			}
			first = Oflag;
		}
		else {
			if (first) {
				for (nsp = sp; !isalnum(*nsp); nsp++)
					continue;
				if (Iflag && isupper(*nsp))
					fp->first = tolower(*nsp);
				else
					fp->first = *nsp;
				fp->pos = *lp;
				fp++;
				first = FALSE;
			}
			fputs(sp, outf);
		}
	} while (sp != NULL);

	/*
	 * write the tables in
	 */

	(void) fclose(inf);
	Tbl.str_numstr = curseek - 1;

	if (Oflag)
		do_order(seekpts, outf);
	else if (Rflag)
		randomize(seekpts);

	(void) fseek(outf, (off_t) 0, 0);
	(void) fwrite((char *) &Tbl, sizeof Tbl, 1, outf);
	(void) fwrite((char *) seekpts, sizeof *seekpts, (int) curseek, outf);
	(void) fclose(outf);

	if (!Sflag) {
		printf("\"%s\" converted to \"%s\"\n", Infile, Outfile);
		if (curseek == 0)
			puts("There was 1 string");
		else
			printf("There were %u strings\n", curseek - 1);
		printf("Longest string: %u byte%s\n", Tbl.str_longlen,
		       Tbl.str_longlen == 1 ? "" : "s");
		printf("Shortest string: %u byte%s\n", Tbl.str_shortlen,
		       Tbl.str_shortlen == 1 ? "" : "s");
	}
	exit(0);
}

/*
 *	This routine evaluates arguments from the command line
 */
getargs(ac, av)
register int	ac;
register char	**av;
{
	register char	*sp;
	register int	i;
	register int	bad, j;

	bad = 0;
	for (i = 1; i < ac; i++)
		if (*av[i] == '-' && av[i][1]) {
			for (sp = &av[i][1]; *sp; sp++)
				switch (*sp) {
				  case 'c': /* new delimiting char */
					if ((Delimch = *++sp) == '\0') {
						--sp;
						Delimch = *av[++i];
					}
					if (Delimch <= 0 || Delimch > '~' ||
					    Delimch == DELIM_CH) {
						printf("bad delimiting character: '\\%o\n'",
						       Delimch);
						bad++;
					}
					break;
				  case 's':	/* silent */
					Sflag++;
					break;
				  case 'v':	/* verbose */
					Sflag = 0;
					break;
				  case 'o':	/* order strings */
					Oflag++;
					break;
				  case 'i':	/* ignore case in ordering */
					Iflag++;
					break;
				  case 'r':	/* ignore case in ordering */
					Rflag++;
					break;
				  default:	/* unknown flag */
					bad++;
					printf("bad flag: '%c'\n", *sp);
					break;
				}
		}
		else if (*av[i] == '-') {
			for (j = 0; Usage[j]; j++)
				puts(Usage[j]);
			exit(0);
		}
		else if (Infile)
			(void) strcpy(Outfile, av[i]);
		else
			Infile = av[i];
	if (!Infile) {
		bad++;
		puts("No input file name");
	}
	if (*Outfile == '\0' && !bad) {
		(void) strcpy(Outfile, Infile);
		(void) strcat(Outfile, ".dat");
	}
	if (bad) {
		puts("use \"strfile -\" to get usage");
		exit(-1);
	}
}

/*
 * do_order:
 *	Order the strings alphabetically (possibly ignoring case).
 */
do_order(seekpts, outf)
off_t	*seekpts;
FILE	*outf;
{
	register int	i;
	register off_t	*lp;
	register STR	*fp;
	extern int	cmp_str();

	(void) fflush(outf);
	Sort_1 = fopen(Outfile, "r");
	Sort_2 = fopen(Outfile, "r");
	Seekpts = seekpts;
	qsort((char *) Firstch, (int) Tbl.str_numstr, sizeof *Firstch, cmp_str);
	i = Tbl.str_numstr;
	lp = seekpts;
	fp = Firstch;
	while (i--)
		*lp++ = fp++->pos;
	(void) fclose(Sort_1);
	(void) fclose(Sort_2);
	Tbl.str_flags |= STR_ORDERED;
}

/*
 * cmp_str:
 *	Compare two strings in the file
 */
cmp_str(p1, p2)
STR	*p1, *p2;
{
	register int	c1, c2;

	c1 = p1->first;
	c2 = p2->first;
	if (c1 != c2)
		return c1 - c2;

	(void) fseek(Sort_1, p1->pos, 0);
	(void) fseek(Sort_2, p2->pos, 0);

	while (!isalnum(c1 = getc(Sort_1)) && c1 != '\0')
		continue;
	while (!isalnum(c2 = getc(Sort_2)) && c2 != '\0')
		continue;

	while (c1 != '\0' && c2 != '\0') {
		if (Iflag) {
			if (isupper(c1))
				c1 = tolower(c1);
			if (isupper(c2))
				c2 = tolower(c2);
		}
		if (c1 != c2)
			return c1 - c2;
		c1 = getc(Sort_1);
		c2 = getc(Sort_2);
	}
	return c1 - c2;
}

/*
 * randomize:
 *	Randomize the order of the string table.  We must be careful
 *	not to randomize across delimiter boundaries.  All
 *	randomization is done within each block.
 */
randomize(seekpts)
register off_t	*seekpts;
{
	register int	cnt, i, j, start;
	register off_t	tmp;
	register off_t	*origsp;

	Tbl.str_flags |= STR_RANDOM;
	srnd(time((time_t *) NULL) + getpid());
	origsp = seekpts;
	for (j = 0; j <= Delim; j++) {

		/*
		 * get the starting place for the block
		 */

		if (j == 0)
			start = 0;
		else
			start = Tbl.str_delims[j - 1];

		/*
		 * get the ending point
		 */

		if (j == Delim)
			cnt = Tbl.str_numstr;
		else
			cnt = Tbl.str_delims[j];

		/*
		 * move things around randomly
		 */

		for (seekpts = &origsp[start]; cnt > start; cnt--, seekpts++) {
			i = rnd(cnt - start);
			tmp = seekpts[0];
			seekpts[0] = seekpts[i];
			seekpts[i] = tmp;
		}
	}
}
